### ISSUE TITLE
Provider stops syncing after iOS wakes

### ISSUE THREAD

**Issue by garth** [ueberdosis/hocuspocus#661]
**Description**
The provider stops receiving updates after iOS wakes from sleep.

**Steps to reproduce the bug**
1. Open a page in Safari on iOS
2. Check that you receive updates
3. Wait for the device to sleep
4. Wake the device
5. Updates are no longer received until the page is reloaded

**Expected behavior**
When the device wakes, if the connection was lost it should be re-established and re-synced

**Environment?**
 - operating system: iOS 16.6
 - browser: Safari
 - mobile/desktop: Mobile
 - Hocuspocus version: @hocuspocus/provider v2.2.3

**Comment by janthurau** [comment]
Hi @garth, thanks for reporting this!

Unfortunately, I cannot reproduce it. Can you reproduce it here? https://tiptap.dev/editor

I opened the page on an iPad & Mac, set the iPad "lock screen" timeout to 2 minutes. After a bit more than 2 minutes, I saw the total number of connections reduce by one. When waking the iPad (after a few more minutes), it immediately (within 2-3 seconds) reconnected and synced again, both incoming and outgoing updates. I've also tried just manually locking the iPad, waiting, and then unlocking again - also worked.

I'm on 16.4 atm, will try to upgrade later and then check again..

**Comment by garth** [comment]
I can confirm that https://tiptap.dev/editor loaded on an iPhone exhibits the same issue. After wake it no longer receives updates.

I'm pretty sure that Safari on iPad is less aggressive with resources than Safari on iPhone.

You don't need to wait for the device screen lock to timeout, you can manually lock the screen with the same effect.

I ran the Safari remote debugger and noticed that after wake there would be a console error about the web socketing being disconnected, so I added the following workaround which fixes the issue:

```ts
    provider.onClose = () => {
      setTimeout(() => {
        void provider.connect()
      }, 100)
    }
```

I not sure if the timeout is necessary, but it doesn't harm.

**Comment by groot9325** [comment]
There was two reason for this case:
1. [WebSocket in latest Safari does not emit onclose event when internet is turned off (wifi or cellular)](https://bugs.webkit.org/show_bug.cgi?id=247943). And I test iOS15/16's WKWebView can stable reproduce. so it cannot trigger re-connect in HocuspocusProviderWebsocket. 
Maybe we show emit close event by manual on iOS which version >= 15 for a workaround way.

2. setTimeout & setInterval on iOS15/16(safari or WKWebView) was not run correctly as expect,maybe for power saving.
   so it caused `checkConnection()` cannnot call periodicly. 
   We can use rAF(requestAnimationFrame) to simulate setInterval (or setTimeout), this npm package wrapper that`@essentials/request-interval`

**Comment by janthurau** [comment]
weird, can't reproduce it on ios17 public beta (iPhone 14 Pro). Thanks for providing your workaround!

Will check how I can reproduce this..

**Comment by janthurau** [comment]
managed to reproduce it, indeed some weird bug causes onClose not being called. Workaround is underway...

**Comment by garth** [comment]
@janthurau In the latest iOS v16, I'm consistently getting a socket closed error after wake. After adding the workaround the connection seems to be very stable. Maybe the issue you are observing is related to iOS v17?

**Comment by groot9325** [comment]
I think if the `setInterval` in iOS16 will run as expect. even if the os sleep and wake, it will be work fine.

the following config can stable re-procude`setInterval` was unreliable in iOS15/iOS16.

1. Config forceSyncInterval 
2. Add debug logs with timestamp in `forceSync()` and `checkConnection()`

the logs timestamp was not correctly print as expect.
```typescript
const socket = new HocuspocusProviderWebsocket({
    url: serverUrl + `${userRoomId}`,
    messageReconnectTimeout:18000, 
    checkConnectionInterval: 6000, // a private config attribute in my forked provider branch
  })

  // Create a websocket provider
  const provider = new HocuspocusProvider(
    {
      websocketProvider: socket,
      token,
      name: `${userRoomId}`,
      document: doc,
      forceSyncInterval: 5000, 
      broadcast: false,
    }
  );
```

iOS15/16 Safari or Webkit( WKWebView) maybe not update by Apple.
So use `requestAnimationFrame` to replace setTimeout in all platform maybe a practical solution.

Of course, consider old  browser or webkit mising `requestAnimationFrame` function, we should still use setTimeout/,fortunately `@essentials/request-interval` has do that.

https://github.com/jaredLunde/essentials/blob/1853a4d8eade1aadee7ac9e10c78e817a4972ebf/packages/raf/src/index.ts#L14C20-L14C20
```
let u = 'undefined',
  win = typeof window !== u ? window : ({} as any),
  p = typeof performance !== u ? performance : Date,
  now = () => p.now(),
  af = 'AnimationFrame' as const,
  Caf = 'cancel' + af,
  Raf = 'request' + af,
  raf: Raf = win[Raf] && win[Raf].bind(win),
  caf: Caf = win[Caf] && win[Caf].bind(win)

if (!raf || !caf) {
  let lastTime = 0
  raf = (callback: FrameRequestCallback) => {
    let curr = now(),
      next = Math.max(lastTime + 1000 / 60, curr)
    return setTimeout(() => {
      callback((lastTime = next))
    }, next - curr) as any
  }
  caf = (h) => clearTimeout(h)
}
```

**Comment by janthurau** [comment]
@garth I've just released v2.3.1 - can you check if that solves your issue? It will take some time, but eventually the connection should reset and fully establish again. If that works in general, I'll see if we can somehow detect the broken socket in a faster way (depending on the configured timeouts, with the default it should take like 20 seconds now).
